testplan/common/utils/graph.py

```diff
@@ -0,0 +1,190 @@
+
+    @classmethod
+    def from_vertices(cls, vertices: Dict[T, U]) -> Self:
+        return cls(
+            copy(vertices),
+            {k: dict() for k in vertices},
+            {k: 0 for k in vertices},
+            {k: 0 for k in vertices},
+        )
+
+    @classmethod
+    def from_vertices_and_edges(
+        cls, vertices: Dict[T, U], edges: Dict[Tuple[T, T], V]
+    ) -> Self:
+        edges_ = {k: dict() for k in vertices}
+        indegrees_ = {k: 0 for k in vertices}
+        outdegrees_ = {k: 0 for k in vertices}
+        for (src, dst), val in edges.items():
+            edges_[src][dst] = val
+            indegrees_[dst] += 1
+            outdegrees_[src] += 1
+
+        return cls(copy(vertices), edges_, indegrees_, outdegrees_)
+
+    def __invalid(self):
+        self.cycles.cache_clear()
+
+    def add_vertex(self, rep: T, obj: U) -> bool:
+        if rep in self.vertices:
+            return False
+        self.__invalid()
+        self.vertices[rep] = obj
+        self.edges[rep] = dict()
+        self.indegrees[rep] = 0
+        self.outdegrees[rep] = 0
+        return True
+
+    def add_edge(self, src: T, dst: T, edge_val: V) -> bool:
+        if (
+            src not in self.edges
+            or dst not in self.edges
+            or dst in self.edges[src]
+        ):
+            return False
+        self.__invalid()
+        self.edges[src][dst] = edge_val
+        self.indegrees[dst] += 1
+        self.outdegrees[src] += 1
+        return True
+
+    def remove_vertex(self, rep: T) -> bool:
+        if rep not in self.vertices:
+            return False
+
+        self.__invalid()
+        del self.vertices[rep]
+        del self.indegrees[rep]
+        del self.outdegrees[rep]
+
+        for v in self.edges[rep]:
+            self.indegrees[v] -= 1
+        del self.edges[rep]
+
+        for v in self.vertices:
+            if rep in self.edges[v]:
+                del self.edges[v][rep]
+                self.outdegrees[v] -= 1
+        return True
+
+    def remove_edge(self, src: T, dst: T) -> bool:
+        if (
+            src not in self.edges
+            or dst not in self.edges
+            or dst not in self.edges[src]
+        ):
+            return False
+
+        self.__invalid()
+        del self.edges[src][dst]
+        self.outdegrees[src] -= 1
+        self.indegrees[dst] -= 1
+        return True
+
+    def update_vertex(self, rep: T, new_obj: U) -> bool:
+        if not rep in self.vertices:
+            return False
+        self.vertices[rep] = new_obj
+        return True
+
+    def update_edge(self, src: T, dst: T, new_val: V) -> bool:
+        if (
+            src not in self.edges
+            or dst not in self.edges
+            or dst not in self.edges[src]
+        ):
+            return False
+        self.edges[src][dst] = new_val
+        return True
+
+    def tarjan_scc(self) -> Dict[T, GraphComponent[T]]:
+        """Tarjan's strongly connected components algorithm"""
```

Comment:
Can we please add a reference for the actual algorithm? https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm is a good first I think, given the references therein, but if you found some freely accessible material, that can help as well (maybe riskier because it eventually can fade, Wiki seems a stable site in existence terms).

Reply:
CiteSeerX perhaps https://citeseerx.ist.psu.edu/doc/10.1.1.327.8418